File: frontend/src/lib/api/students.ts
import { BaseApiService } from './base';
import { API_ENDPOINTS } from './types';
import type { ArchiveResponse } from './archive';

// Eligibility interfaces
export interface EligibilityCategory {
  id: number;
  name: string;
  code?: string;
  description?: string;
  is_active: boolean;
  display_order?: number;
  created_date: string;
  modified_date: string;
}

export interface StudentEligibility {
  id: number;
  student_id: number;
  eligibility_category_id: number;
  start_date: string;
  end_date?: string;
  is_primary: boolean;
  notes?: string;
  created_date: string;
  modified_date: string;
  eligibility_category: EligibilityCategory;
  is_active: boolean;
}

// Teacher summary interface for student relationships
export interface TeacherSummary {
  id: number;
  first_name: string;
  last_name: string;
  title?: string;
  email?: string;
  display_name: string;
  full_name: string;
}

// Student data interfaces
export interface Student {
  id: number;
  first: string;
  last: string;
  uic?: string;
  grade_level?: string;
  enrollment_status: string;
  is_archived: boolean;
  date_of_birth?: string;
  created_date: string;
  modified_date: string;
  // IEP and Evaluation dates
  iep_date?: string;
  annual_review_due_date?: string;
  reevaluation_due_date?: string;
  iep_meeting_date?: string;
  initial_evaluation_date?: string;
  eligibility_determination_date?: string;
  // School and teacher assignments
  school_id?: number;
  teacher_id?: number;
  case_manager_id?: number;
  teacher?: TeacherSummary;
  case_manager?: TeacherSummary;
  school?: { id: number; name: string; district?: string };
  // Eligibility information
  eligibilities: StudentEligibility[];
}

export interface StudentSummary {
  id: number;
  first: string;
  last: string;
  grade_level?: string;
  case_manager?: string | TeacherSummary;  // Can be legacy string or new object
  enrollment_status: string;
  is_archived: boolean;
  school_id?: number;  // Added for scheduling functionality
  uic?: string;  // Added for student identification
  teacher_id?: number;
  case_manager_id?: number;
  teacher?: TeacherSummary;
}

export interface CreateStudentRequest {
  first: string;
  last: string;
  uic?: string;
  grade_level?: string;
  teacher_id?: number;
  case_manager_id?: number;
  enrollment_status?: string;
  is_archived?: boolean;
  date_of_birth?: string;
  // IEP and Evaluation dates
  iep_date?: string;
  annual_review_due_date?: string;
  reevaluation_due_date?: string;
  iep_meeting_date?: string;
  initial_evaluation_date?: string;
  eligibility_determination_date?: string;
  school_id?: number;
}

export interface UpdateStudentRequest {
  first?: string;
  last?: string;
  uic?: string;
  grade_level?: string;
  enrollment_status?: string;
  is_archived?: boolean;
  date_of_birth?: string;
  // IEP and Evaluation dates
  iep_date?: string;
  annual_review_due_date?: string;
  reevaluation_due_date?: string;
  iep_meeting_date?: string;
  initial_evaluation_date?: string;
  eligibility_determination_date?: string;
  // School and teacher assignments
  school_id?: number | null;
  teacher_id?: number | null;
  case_manager_id?: number | null;
}

export interface StudentsFilters {
  enrollment_status?: string;
  case_manager?: string;
  include_archived?: boolean;
}

class StudentsApiService extends BaseApiService {
  constructor() {
    super(API_ENDPOINTS.STUDENTS);
  }

  // List students with optional filtering
  async getStudents(filters?: StudentsFilters): Promise<StudentSummary[]> {
    const params: Record<string, any> = {};
    
    if (filters?.enrollment_status) {
      params.enrollment_status = filters.enrollment_status;
    }
    if (filters?.case_manager) {
      params.case_manager = filters.case_manager;
    }
    if (filters?.include_archived !== undefined) {
      params.include_archived = filters.include_archived;
    }
    
    return this.get<StudentSummary[]>('', params);
  }

  // Get a specific student by ID
  async getStudent(id: number): Promise<Student> {
    return this.get<Student>(`/${id}`);
  }

  // Get a student by UIC (for legacy integration)
  async getStudentByUic(uic: string): Promise<Student> {
    return this.get<Student>(`/uic/${uic}`);
  }

  // Create a new student
  async createStudent(data: CreateStudentRequest): Promise<Student> {
    return this.post<Student>('', data);
  }

  // Update an existing student
  async updateStudent(id: number, data: UpdateStudentRequest): Promise<Student> {
    return this.put<Student>(`/${id}`, data);
  }

  // ARCHIVES the student and everything under them. Same verb, same path --
  // the backend stopped deleting; see backend/app/routers/students.py. The
  // response carries `archiveEventId`, which is what makes an undo possible.
  async deleteStudent(id: number): Promise<ArchiveResponse> {
    return this.delete<ArchiveResponse>(`/${id}`);
  }

  // Archive a student (hide from active lists but preserve data)
  async archiveStudent(id: number): Promise<Student> {
    return this.put<Student>(`/${id}/archive`, {});
  }

  // Unarchive a student (restore to active lists)
  async unarchiveStudent(id: number): Promise<Student> {
    return this.put<Student>(`/${id}/unarchive`, {});
  }

  // Get archived students
  async getArchivedStudents(): Promise<StudentSummary[]> {
    return this.get<StudentSummary[]>('/archived');
  }

  // Bulk operations (future enhancement)
  async bulkCreateStudents(students: CreateStudentRequest[]): Promise<Student[]> {
    return this.post<Student[]>('/bulk', { students });
  }

  // Get students by case manager (convenience method)
  async getStudentsByCaseManager(caseManager: string): Promise<StudentSummary[]> {
    return this.getStudents({ case_manager: caseManager });
  }

  // Get active students only (convenience method)
  async getActiveStudents(): Promise<StudentSummary[]> {
    return this.getStudents({ enrollment_status: 'Active' });
  }
}

// Export singleton instance
export const studentsApi = new StudentsApiService();

// Export for backward compatibility
export default studentsApi;

File: deploy_temp/app/routers/students.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional

from app.db.database import get_db
from app.repositories.student_repository import StudentRepository
from app.schemas.student import StudentCreate, StudentRead, StudentUpdate, StudentSummary


router = APIRouter(prefix="/api/students", tags=["students"])


@router.get("/", response_model=List[StudentSummary])
@router.get("", response_model=List[StudentSummary])  # Handle both with and without trailing slash
def list_students(
    enrollment_status: Optional[str] = Query(None, description="Filter by enrollment status"),
    case_manager: Optional[str] = Query(None, description="Filter by case manager"),
    include_archived: bool = Query(False, description="Include archived students in results"),
    db: Session = Depends(get_db)
):
    """List students with optional filtering"""
    repo = StudentRepository(db)
    
    if case_manager:
        return repo.get_students_by_case_manager(case_manager, include_archived=include_archived)
    else:
        return repo.list_students(enrollment_status=enrollment_status, include_archived=include_archived)


@router.get("/archived", response_model=List[StudentSummary])
def get_archived_students(db: Session = Depends(get_db)):
    """Get all archived students"""
    repo = StudentRepository(db)
    return repo.get_archived_students()


@router.get("/{student_id}", response_model=StudentRead)
def get_student(student_id: int, db: Session = Depends(get_db)):
    """Get a specific student by ID"""
    repo = StudentRepository(db)
    student = repo.get_student_by_id(student_id)
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    
    return student


@router.post("/", response_model=StudentRead)
@router.post("", response_model=StudentRead)  # Handle both with and without trailing slash
def create_student(payload: StudentCreate, db: Session = Depends(get_db)):
    """Create a new student"""
    repo = StudentRepository(db)
    
    # Check if UIC already exists (if provided)
    if payload.uic:
        existing = repo.get_student_by_uic(payload.uic)
        if existing:
            raise HTTPException(status_code=400, detail="UIC already exists")
    
    return repo.create_student(payload)


@router.put("/{student_id}", response_model=StudentRead)
def update_student(student_id: int, payload: StudentUpdate, db: Session = Depends(get_db)):
    """Update an existing student"""
    repo = StudentRepository(db)
    
    # Check if UIC conflicts (if being updated)
    if payload.uic:
        existing = repo.get_student_by_uic(payload.uic)
        if existing and existing.id != student_id:
            raise HTTPException(status_code=400, detail="UIC already exists")
    
    student = repo.update_student(student_id, payload)
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    return student


@router.delete("/{student_id}")
def delete_student(student_id: int, db: Session = Depends(get_db)):
    """Delete a student (permanent removal)"""
    repo = StudentRepository(db)
    success = repo.delete_student(student_id)
    if not success:
        raise HTTPException(status_code=404, detail="Student not found")
    return {"message": "Student deleted successfully"}


@router.put("/{student_id}/archive", response_model=StudentRead)
def archive_student(student_id: int, db: Session = Depends(get_db)):
    """Archive a student (hide from active lists but preserve data)"""
    repo = StudentRepository(db)
    
    student = repo.archive_student(student_id)
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    
    return student


@router.put("/{student_id}/unarchive", response_model=StudentRead)
def unarchive_student(student_id: int, db: Session = Depends(get_db)):
    """Unarchive a student (restore to active lists)"""
    repo = StudentRepository(db)
    
    student = repo.unarchive_student(student_id)
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    
    return student


@router.get("/uic/{uic}", response_model=StudentRead)
def get_student_by_uic(uic: str, db: Session = Depends(get_db)):
    """Get a student by their UIC (for legacy system integration)"""
    repo = StudentRepository(db)
    student = repo.get_student_by_uic(uic)
    if not student:
        raise HTTPException(status_code=404, detail="Student not found")
    return student


